group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -936,62 +941,90 @@ private void append(
-                            // deduce it like this.
+                    for (int i = 0; i < records.size(); i++) {
+                        U recordToReplay = records.get(i);
+                        SimpleRecord recordToAppend = recordsToAppend.get(i);
+
+                        if (replay) {
                             coordinator.replay(
-                                currentBatch.nextOffset + i,
+                                currentBatch.nextOffset,
                                 producerId,
                                 producerEpoch,
-                                records.get(i)
+                                recordToReplay
                             );
                         }
-                    }
 
-                    // Append to the batch.
-                    for (SimpleRecord record : recordsToAppend) {
-                        currentBatch.builder.append(record);
+                        if (!isAtomic) {
+                            boolean hasRoomFor = currentBatch.builder.hasRoomFor(
+                                recordToAppend.timestamp(),
+                                recordToAppend.key(),
+                                recordToAppend.value(),
+                                recordToAppend.headers()
+                            );
+
+                            if (!hasRoomFor) {
+                                if (currentBatch.builder.numRecords() == 0) {
+                                    throw new RecordTooLargeException("Record " + recordToAppend + " in append to partition " + tp +
+                                        " exceeds exceeds the maximum configured size of " + currentBatch.maxBatchSize + ".");
+                                }
+
+                                // If the current batch is not empty, we flush it and allocate a new batch.
+                                flushCurrentBatch();
+                                maybeAllocateNewBatch(
+                                    producerId,
+                                    producerEpoch,
+                                    verificationGuard,
+                                    currentTimeMs
+                                );
+                            }
+                        }
+
+                        currentBatch.builder.append(recordToAppend);
                         currentBatch.nextOffset++;
                     }
+
+                    // Add the event to the list of pending events associated with the batch.
+                    currentBatch.deferredEvents.add(event);
                 } catch (Throwable t) {
                     log.error("Replaying records to {} failed due to: {}.", tp, t.getMessage());
+
+                    // Add the event to the list of pending events associated with the last
+                    // batch in order to fail it too.
+                    currentBatch.deferredEvents.add(event);
```

Comment:
the reason we're duplicating this line (instead of adding the event at the start of the try block) is to prevent future code changes from breaking this logic?

Reply:
Partially. The main reason was that the event must be attached to the last batch (vs the first one) in order to only complete the write when all records are committed.